server/market/balancer.go

```diff
@@ -0,0 +1,113 @@
+}
+
+// MatchNegotiator can view match-reserved funds for an account-based asset's
+// address. MatchNegotiator is satisfied by *Swapper.
+type MatchNegotiator interface {
+	// AccountStats returns the qty, swap count, and redeem count pending in
+	// active matches.
+	AccountStats(acctAddr string, assetID uint32) (qty, swaps uint64, redeems int)
+}
+
+// BackedBalancer is an asset manager that is capable of querying the entire DEX
+// for the balance required to fulfill new + existing orders and outstanding
+// redemptions.
+type DEXBalancer struct {
+	tunnels         map[string]PendingAccounter
+	assets          map[uint32]*backedBalancer
+	matchNegotiator MatchNegotiator
+}
+
+// NewDEXBalancer is a constructor for a DEXBalancer. Provided assets will
+// be filtered for those that are account-based. The matchNegotiator is
+// satisfied by the *Swapper.
+func NewDEXBalancer(tunnels map[string]PendingAccounter, assets map[uint32]*asset.BackedAsset, matchNegotiator MatchNegotiator) *DEXBalancer {
+	balancers := make(map[uint32]*backedBalancer)
+	for assetID, ba := range assets {
+		balancer, is := ba.Backend.(asset.AccountBalancer)
+		if !is {
+			continue
+		}
+		balancers[assetID] = &backedBalancer{
+			balancer:  balancer,
+			assetInfo: &ba.Asset,
+		}
+	}
+
+	return &DEXBalancer{
+		tunnels:         tunnels,
+		assets:          balancers,
+		matchNegotiator: matchNegotiator,
+	}
+}
+
+// CheckBalance checks if there is sufficient balance to support the specified
+// new funding and redemptions, given the existing orders throughout DEX that
+// fund from or redeem to the specified account address for the account-based
+// asset. It is an internally logged error to call CheckBalance for a
+// non-account-based asset or an asset that is was not provided to the
+// constructor.
+func (b *DEXBalancer) CheckBalance(acctAddr string, assetID uint32, qty, lots uint64, redeems int) bool {
+	backedAsset, found := b.assets[assetID]
+	if !found {
+		log.Errorf("(*DEXBalancer).CheckBalance: asset ID %d not a configured backedBalancer", assetID)
+		return false
+	}
+
+	log.Tracef("balance check for %s - %s: new qty = %d, new lots = %d, new redeems = %d",
+		backedAsset.assetInfo.Symbol, acctAddr, qty, lots, redeems)
+
+	bal, err := backedAsset.balancer.AccountBalance(acctAddr)
+	if err != nil {
+		log.Error("(*DEXBalancer).CheckBalance: error getting account balance for %q: %v", acctAddr, err)
+		return false
+	}
+
+	// Add quantity for unfilled orders.
+	for _, mt := range b.tunnels {
+		newQty, newLots, newRedeems := mt.AccountPending(acctAddr, assetID)
+		lots += newLots
+		qty += newQty
+		redeems += newRedeems
+	}
+
+	// Add in-process swaps.
+	newQty, newLots, newRedeems := b.matchNegotiator.AccountStats(acctAddr, assetID)
+	lots += newLots
+	qty += newQty
+	redeems += newRedeems
+
+	assetInfo := backedAsset.assetInfo
+	redeemCosts := uint64(redeems) * assetInfo.RedeemSize * assetInfo.MaxFeeRate
```

Comment:
Since `MaxFeeRate` isn't actually required for redeems, I think this deserves a comment that this is a conservatively high estimate of possible redeem costs.